include/deal.II/base/function_spherical.h

```diff
@@ -0,0 +1,122 @@
+
+namespace Functions
+{
+
+  /**
+   * An abstract base class for a scalar-valued function $f=f(r,\theta,\phi)$
+   * defined in spherical coordinates. This class wraps transformation of values,
+   * gradients and hessians from spherical coordinates to Cartesian coordinate system.
+   * Therefore derived classes only need to implement those functions in
+   * spherical coordinates (specifically svalue(), sgradient() and
+   * shessian() ). The convention for angles is the same as in
+   * GeometricUtilities::Coordinates.
+   *
+   * @note This function is currently only implemented for dim==3 .
+   *
+   * @author Denis Davydov, 2017
+   */
+  template <int dim>
+  class Spherical : public Function<dim>
+  {
+  public:
+    /**
+     * Constructor which should be provided with @p center defining the origin
+     * of the coordinate system.
+     */
+    Spherical(const Point<dim> &center);
+
+    /**
+     * Virtual destructor.
+     */
+    virtual ~Spherical();
+
+    /**
+     * Return the value of the function at the given point.
+     */
+    virtual double value (const Point<dim> &point,
+                          const unsigned int component = 0) const;
+
+    /**
+     * Return the gradient in Cartesian coordinates at point @p p.
+     */
+    virtual Tensor<1,dim> gradient (const Point<dim>   &p,
+                                    const unsigned int  component = 0) const;
+
+    /**
+     * Return the Hessian in Cartesian coordinates at point @p p.
+     */
+    virtual SymmetricTensor<2,dim> hessian (const Point<dim> &p,
+                                            const unsigned int component=0) const;
+
+    std::size_t memory_consumption () const;
+
+  private:
+    /**
+     * Return the value at point @p sp.
+     */
+    virtual double svalue(const std_cxx11::array<double, dim> &sp) const = 0;
+
+    /**
+     * Return the gradient in spherical coordinates.
+     *
+     * The returned object should contain the following components:
+     * $G_i = \frac{df}{dy_i}\quad\rm{where}\quad y_i=\{r,\theta,\phi\}$.
+     */
+    virtual Tensor<1,dim> sgradient(const std_cxx11::array<double, dim> &sp) const = 0;
+
+    /**
+     * Return the Hessian in spherical coordinates.
+     *
+     * The returned object should contain the following components:
+     * $H_{ij} = \frac{d^2f}{dy_idy_j}\quad\rm{where}\quad y_{i,j}=\{r,\theta,\phi\}$.
+     */
+    virtual SymmetricTensor<2,dim> shessian (const std_cxx11::array<double, dim> &sp) const = 0;
+
+    /**
+     * A vector from the origin to the center of spherical coordinate system.
+     */
+    const Tensor<1,dim> coordinate_system_offset;
+
+    /**
+     * Evaluate unit vectors in spherical coordinates
+     */
+    void set_unit_vectors(Tensor<1,dim> &unit_r,
+                          Tensor<1,dim> &unit_theta,
+                          Tensor<1,dim> &unit_phi,
+                          const double &cos_theta,
+                          const double &sin_theta,
```

Comment:
put the input arguments first.

That said, this class has no state (= member variables), so I suspect that the function could be `static`. If so, it may also be moved into an anonymous namespace in the `.cc` file, right?